File: edge/app/discovery.py
from __future__ import annotations

import os
import re
from dataclasses import dataclass
from pathlib import Path

import yaml


UPLOAD_DIR_FILENAME = ".upload_dir"
SAFE_COMPONENT_PATTERN = re.compile(r"^[A-Za-z0-9._-]+$")


@dataclass(slots=True)
class JobDefinition:
    root_path: Path
    job_name: str
    exclude_patterns: list[str]
    include_hidden: bool
    follow_symlinks: bool

    @property
    def state_key(self) -> str:
        return str(self.root_path.resolve())


def discover_jobs(scan_root: Path, max_depth: int, logger) -> list[JobDefinition]:
    jobs: list[JobDefinition] = []

    def walk(directory: Path, depth: int) -> None:
        if depth > max_depth:
            return

        try:
            entries = list(os.scandir(directory))
        except (FileNotFoundError, NotADirectoryError, PermissionError, OSError) as exc:
            logger.warning("skipped_missing path=%s detail=%s", directory, exc)
            return

        marker = next(
            (
                Path(entry.path)
                for entry in entries
                if entry.name == UPLOAD_DIR_FILENAME and entry.is_file(follow_symlinks=False)
            ),
            None,
        )

        if marker is not None:
            job = load_job_definition(directory=directory, marker_path=marker, logger=logger)
            if job is not None:
                jobs.append(job)
                logger.info("job_discovered job_name=%s path=%s", job.job_name, job.root_path)
            return

        if depth == max_depth:
            return

        for entry in entries:
            if not entry.is_dir(follow_symlinks=False):
                continue
            walk(Path(entry.path), depth + 1)

    walk(scan_root, 0)
    return jobs


def load_job_definition(directory: Path, marker_path: Path, logger) -> JobDefinition | None:
    try:
        raw_content = marker_path.read_text(encoding="utf-8")
        payload = yaml.safe_load(raw_content) if raw_content.strip() else {}
    except yaml.YAMLError as exc:
        logger.error("invalid_upload_dir path=%s detail=%s", marker_path, exc)
        return None
    except OSError as exc:
        logger.error("invalid_upload_dir path=%s detail=%s", marker_path, exc)
        return None

    if payload is None:
        payload = {}

    if not isinstance(payload, dict):
        logger.error("invalid_upload_dir path=%s detail=expected mapping", marker_path)
        return None

    job_name = str(payload.get("job_name") or directory.name).strip()
    if not SAFE_COMPONENT_PATTERN.fullmatch(job_name):
        logger.error("invalid_upload_dir path=%s detail=invalid job_name", marker_path)
        return None

    exclude = payload.get("exclude") or []
    if not isinstance(exclude, list) or not all(isinstance(item, str) for item in exclude):
        logger.error("invalid_upload_dir path=%s detail=exclude must be a list of strings", marker_path)
        return None

    include_hidden = payload.get("include_hidden", True)
    follow_symlinks = payload.get("follow_symlinks", False)
    if not isinstance(include_hidden, bool) or not isinstance(follow_symlinks, bool):
        logger.error("invalid_upload_dir path=%s detail=boolean options must be true or false", marker_path)
        return None

    return JobDefinition(
        root_path=directory.resolve(),
        job_name=job_name,
        exclude_patterns=list(exclude),
        include_hidden=include_hidden,
        follow_symlinks=follow_symlinks,
    )

File: edge/internal/backup/discovery.go
package backup

import (
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"slices"
	"strings"

	"gopkg.in/yaml.v3"
)

const UploadDirFilename = ".upload_dir"

var safeComponentPattern = regexp.MustCompile(`^[A-Za-z0-9._-]+$`)

// JobDefinition describes a single backup job discovered from a .upload_dir marker.
type JobDefinition struct {
	RootPath        string
	JobName         string
	ExcludePatterns []string
	IncludeHidden   bool
	FollowSymlinks  bool
}

// StateKey is the filesystem path used to key job state.
func (j *JobDefinition) StateKey() string {
	return j.RootPath
}

// DiscoverJobs walks scanRoot up to maxDepth levels looking for .upload_dir markers.
func DiscoverJobs(scanRoot string, maxDepth int, warnf func(string, ...any)) ([]*JobDefinition, error) {
	type entry struct {
		dir   string
		depth int
	}
	queue := []entry{{scanRoot, 0}}
	var jobs []*JobDefinition

	for len(queue) > 0 {
		cur := queue[0]
		queue = queue[1:]

		if cur.depth > maxDepth {
			continue
		}

		entries, err := os.ReadDir(cur.dir)
		if err != nil {
			warnf("skipped_missing path=%s detail=%s", cur.dir, err)
			continue
		}

		var markerFound bool
		for _, e := range entries {
			if e.Name() == UploadDirFilename && !e.IsDir() {
				markerFound = true
				break
			}
		}

		if markerFound {
			markerPath := filepath.Join(cur.dir, UploadDirFilename)
			job, err := LoadJobDefinition(cur.dir, markerPath, warnf)
			if err == nil && job != nil {
				jobs = append(jobs, job)
			}
			continue
		}

		if cur.depth == maxDepth {
			continue
		}

		var children []string
		for _, e := range entries {
			if e.IsDir() {
				children = append(children, filepath.Join(cur.dir, e.Name()))
			}
		}
		slices.SortFunc(children, func(a, b string) int {
			return strings.Compare(strings.ToLower(filepath.Base(a)), strings.ToLower(filepath.Base(b)))
		})
		for _, child := range children {
			queue = append(queue, entry{child, cur.depth + 1})
		}
	}
	return jobs, nil
}

// LoadJobDefinition reads a .upload_dir marker and returns the job, or nil on error.
func LoadJobDefinition(dir, markerPath string, warnf func(string, ...any)) (*JobDefinition, error) {
	payload, err := ReadUploadDirPayload(markerPath)
	if err != nil {
		if warnf != nil {
			warnf("invalid_upload_dir path=%s detail=%s", markerPath, err)
		}
		return nil, err
	}
	job, err := BuildJobDefinition(dir, payload)
	if err != nil {
		if warnf != nil {
			warnf("invalid_upload_dir path=%s detail=%s", markerPath, err)
		}
		return nil, err
	}
	return job, nil
}

// ReadUploadDirPayload reads and parses the YAML content of a .upload_dir file.
func ReadUploadDirPayload(markerPath string) (map[string]any, error) {
	raw, err := os.ReadFile(markerPath)
	if err != nil {
		return nil, err
	}
	content := strings.TrimSpace(string(raw))
	if content == "" {
		return map[string]any{}, nil
	}
	var payload map[string]any
	if err := yaml.Unmarshal(raw, &payload); err != nil {
		return nil, fmt.Errorf("invalid yaml: %w", err)
	}
	if payload == nil {
		return map[string]any{}, nil
	}
	return payload, nil
}

// BuildJobDefinition constructs a JobDefinition from a directory path and YAML payload.
func BuildJobDefinition(dir string, payload map[string]any) (*JobDefinition, error) {
	absDir, err := filepath.Abs(dir)
	if err != nil {
		return nil, err
	}

	jobName := strings.TrimSpace(stringVal(payload["job_name"], filepath.Base(absDir)))
	if !safeComponentPattern.MatchString(jobName) {
		return nil, fmt.Errorf("invalid job_name")
	}

	excludeRaw, _ := payload["exclude"]
	excludePatterns, err := parseStringList(excludeRaw)
	if err != nil {
		return nil, fmt.Errorf("exclude must be a list of strings")
	}

	includeHidden := boolVal(payload["include_hidden"], true)
	followSymlinks := boolVal(payload["follow_symlinks"], false)

	return &JobDefinition{
		RootPath:        absDir,
		JobName:         jobName,
		ExcludePatterns: excludePatterns,
		IncludeHidden:   includeHidden,
		FollowSymlinks:  followSymlinks,
	}, nil
}

// JobDefinitionToPayload converts a JobDefinition to a serializable map.
func JobDefinitionToPayload(job *JobDefinition) map[string]any {
	exclude := make([]string, len(job.ExcludePatterns))
	copy(exclude, job.ExcludePatterns)
	return map[string]any{
		"job_name":        job.JobName,
		"exclude":         exclude,
		"include_hidden":  job.IncludeHidden,
		"follow_symlinks": job.FollowSymlinks,
	}
}

// WriteUploadDir serializes a payload and writes it to directory/.upload_dir.
func WriteUploadDir(dir string, payload map[string]any) error {
	job, err := BuildJobDefinition(dir, payload)
	if err != nil {
		return err
	}
	normalized := JobDefinitionToPayload(job)
	out, err := yaml.Marshal(normalized)
	if err != nil {
		return err
	}
	return os.WriteFile(filepath.Join(dir, UploadDirFilename), out, 0o644)
}

// DeleteUploadDir removes the .upload_dir marker from a directory.
func DeleteUploadDir(dir string) error {
	path := filepath.Join(dir, UploadDirFilename)
	if err := os.Remove(path); err != nil && !os.IsNotExist(err) {
		return err
	}
	return nil
}

func stringVal(v any, def string) string {
	if v == nil {
		return def
	}
	s := strings.TrimSpace(fmt.Sprintf("%v", v))
	if s == "" {
		return def
	}
	return s
}

func boolVal(v any, def bool) bool {
	if v == nil {
		return def
	}
	if b, ok := v.(bool); ok {
		return b
	}
	return def
}

func parseStringList(v any) ([]string, error) {
	if v == nil {
		return nil, nil
	}
	list, ok := v.([]any)
	if !ok {
		return nil, fmt.Errorf("not a list")
	}
	result := make([]string, 0, len(list))
	for _, item := range list {
		s, ok := item.(string)
		if !ok {
			return nil, fmt.Errorf("list contains non-string item")
		}
		result = append(result, s)
	}
	return result, nil
}
